include/rdma/fabric.h

```diff
@@ -316,6 +316,21 @@ enum {
 	FI_PROTO_EFA
 };
 
+#define FI_TC_PROV_SPECIFIC (1UL << 31)
+
+enum {
+	FI_TC_UNSPEC = 0,
+	FI_TC_DSCP = FI_TC_UNSPEC,
```

Comment:
I would not have DSCP the same value as UNSPEC.  We want a way for the app to set the TC for the domain, and have that value be used as the default.  This is only possible if the EP or TX TC can both be set to UNSPEC.

Otherwise, if the domain TC is set to, say, BEST_EFFORT, but the EP/TX are set to UNSPEC.  Then the EP/TX will end up using DSCP, because we can't distinguish between that and UNSPEC.

Reply:
I'll just remove the DSCP label and add some documentation to the man page to make it clear that any value other than 0, under 0x100 is a DSCP value.

Reply:
Actually, we can't really do that. In order to support the best effort DSCP value, we need 0 to be accessible.

I see two options so far:

- provide an offset for DSCP and require providers to inspect the traffic class value to determine if it is a label, a DSCP value, or unspec. If DSCP, they need to remove the offset to determine what DSCP value the client wants to use. if the offset were 0x100, then using the best effort DSCP value would be (0x100 + 0x00). That's going to make the use of DSCP a little clunky. 
- make the use of FI_TC_UNSPEC a conscious decision on the part of the application, or fill in the tclass value when an fi_info structure is allocated via fi_allocinfo. To make it a conscious decision, FI_TC_UNSPEC would have to be non-zero. This has the benefit of making the use of DSCP feel seamless, since there is no offset manipulation. We could advise providers to ignore the tclass field in fi_info structures where the libfabric version requested by the application was less than 1.9/1.3

Reply:
I definitely want to keep UNSPEC = 0.  That's consistent with the rest of the API and falls out naturally from calling fi_allocinfo().  A DSCP flag may be the best option here.  (It's equivalent to your offset mode).  This kind of goes back to your original proposal of having tclass be interpreted as fields...  :-/

We can make it easier on the user if we provide some helper macros.  e.g.
```
#define FI_TC_DSCP(x) (0x10000000 | (x))

ep_attr.tclass = FI_TC_DSCP(0)
```

Reply:
That can work for me.